### DISCUSSION TITLE
Multithreaded Rendering?

### DISCUSSION THREAD

**Discussion by IAmAThousandTrees** [ManimCommunity/manim#3897] [Ideas]
I'm not too familiar with the internals of Manim or how python does threading, or even python if I'm being honest (I'm really only learning python in order to use Manim), so there might be something preventing this from being viable that I cannot see... but it seems to me that there's a big missed opportunity in not threading the rendering process.

I'm making video's that have hundreds of animations in a 10-minute run-time and it takes upwards of 30 minutes to even re-render (with 90% of clips only being hashed) after a small edit.

My thought is this: wouldn't the hashing/clip-re-use process provide an ideal junction point for threading the process? everything required to uniquely specify an animation is gathered together for the process of hashing it. If all of that gets packaged up and handed to a worker thread as its task (deep object copy?) - to hash and then render if required - then all that is required is to simulate the results of the animation, which i presume is done anyway if the hash finds the render unnecessary, and continue with running the user code in the main thread.

Wait for all the render threads to return the files they've identified/rendered, smoosh them together as usual, done. 

It's probably sensible to limit the number of concurrent threads to the N of CPU threads by blocking the main thread waiting on any worker thread return when starting the Nth thread.

The progress reporting might have to be re-written for handling multiple simultaneous renders, but that's not a big issue.

**Comment by JasonGrace2282** [comment]
I think the developers have had lots of conversations about this - if you have any questions, feel free to ping us on discord :)
Before I go any further, I do just want to note that multithreading would go against the idea of live rendering with OpenGL.

The problem is really in two parts - the current architecture of Manim is insufficient to handle multithreading correctly, and the user has an incredible amount of "power" in what they can do. I'll tackle the second point first, because it's easier to describe.

Take this code for example:
```py
sq = Square()
self.add(sq)
self.play(sq.animate.shift(RIGHT))
self.play(FadeOut(sq))
```
It's pretty obvious that the way the second animation (`FadeOut`) looks depends on the state after the first animation. This can (probably) be worked around, by abusing the way animations work. However, there are more complicated examples (typically related to updaters - see [`TracedPath`](https://docs.manim.community/en/stable/reference/manim.animation.changing.TracedPath.html)) where the final state of an object depends on each frame throughout each animation. There's no way to figure out the final state of the object in advance, and so there's no way we can dispatch a worker thread for the next `play` call without finishing the current `play` call.

Now, the next thought is to make the user explicitly opt-in to threading their code (maybe checking if updaters exist, or forcing users to mark updaters with an attribute if they are "pure"). Trying to handle this with our current architecture ([relevant docs](https://docs.manim.community/en/stable/guides/deep_dive.html#the-play-call-preparing-to-enter-manim-s-render-loop)) is an absolute nightmare, and is bound to cause bugs. In order to do this in the least hacky way possible, we would have to rewrite the `Scene` to use a two pass rendering system, where the first pass runs through the user code and gathers all the actions that need to happen, and the second pass executes them (threading them where possible). As you can imagine, this is a HUGE change that would need to happen.

But now! All hope is not lost! Unlike what it may seem like, the devs are still making huge leaps towards improving Manim. We have a refactor branch where we've rewritten the renderer and `Scene` by introducing a `Manager` class to handle all the rendering logistics, which should make it much easier to possibly parallelize certain animations. I wouldn't expect it to be ready for any time before Manim v0.21.0, but if you want to check out the progress, you can always take a look at #3817

**Reply by behackl** [reply]
A minor addition to this answer, just to make one part more explicit: if it were the case that the hashes were computed by simulating the result of animations and then hashing the entire scene, then you would actually be correct -- in that case, slicing up the rendering process and dispatching the segments to various workers would be feasible (although we'd have to be careful about the amount of object copies being created and passed to workers).

However, this is unfortunately not how things are currently implemented. The hash is computed by combining a hash of the current state of the scene (before the animation is applied) with a hash of the animation(s) to be applied. Manim does not know how the result of the animation looks like when computing the hashes used to identify the cache segments.

It will require the rewrite mentioned by Jason to change how this is handled. Help with any part of the process would be much appreciated!

**Reply by IAmAThousandTrees** [reply]
So, in attempting to understand how the animation-hashing and animation-skipping options seem to defy your assertion that skipping animating an animation isn't possible, I went to look at the cairo-renderer's play function and then to the called-back scene.play_internal function, and it seems to me that what I proposed is entirely possible.

The cairo_renderer.play function checks for whether the animation is supposed to be skipped, then checks if the hashing function returns a known hash, updates the skipping status, calls to scene.play_internal to perform the animation (skipped or unskipped) and then adds the animation to the list of included animation hashes for inclusion in the final result.

The scene.play_internal function iterates over the frames of the animation, performing all frame-by-frame updates to the scenes objects, and then _if the animation is not being skipped_ renders the frame.

In this process there is definitely room for multithreading.

If in the renderer.play function, in the same "if" that checks the hash of the animation, one were to instead commission a worker thread with the scene as one of its parameters running, a function that continues the rendering from that point (hash, call to scene.play_internal, etc, log the rendered animation hash), then continue the main thread with skip_animations set to true so that the main thread doesn't render the animation, just simulates and continues.

Researching a little about how python limits threading options and how people get around that, it looks like using the mutiprocessing module to spawn a set of worker processes and its queue and priority_queue classes to send sequence-tagged scenes and receive sequence-tagged animation hashes back in order, together with a loop at the end to receive all the hashes which will block until all expected animations are returned, might be an effective method.

The only limitations on whether this is doable would be if everything required for the hashing and rendering of the animations isn't included in the scene object tree (the scene contains the mobjects, renderer, and animations ... is there anything else?). Since the workers are separate processes, the scene object tree would be serialized in its entirety anyway when sending to the multiprocess queue, so a deep_copy isn't needed.

The simulation of the animations seems to be quick (from having used the skip-animations command-line option: skipping 100 animations takes a few seconds compared to several minutes to hash them) so the limitation on whether this is worthwhile will hinge on the overhead of the serialization used by multiprocessing.queue, which I presume is pickling, so should also be quite quick.

I also just noticed that the hashing is done via a hand-rolled json serializer and I wonder (since this is actually what is taking all the time in my case) if simply pickling the objects and crc32-ing _that_ wouldn't be much faster?

**Reply by behackl** [reply]
> If in the renderer.play function, in the same "if" that checks the hash of the animation, one were to instead commission a worker thread with the scene as one of its parameters running, a function that continues the rendering from that point (hash, call to scene.play_internal, etc, log the rendered animation hash), then continue the main thread with skip_animations set to true so that the main thread doesn't render the animation, just simulates and continues.

Sounds reasonable to me. 

> The only limitations on whether this is doable would be if everything required for the hashing and rendering of the animations isn't included in the scene object tree (the scene contains the mobjects, renderer, and animations ... is there anything else?).

Not as far as I am aware of, at least there isn't really anything else that is being hashed.

> I also just noticed that the hashing is done via a hand-rolled json serializer and I wonder (since this is actually what is taking all the time in my case) if simply pickling the objects and crc32-ing that wouldn't be much faster?

One problem with pickling (which is only somewhat possible since the color rework) is updater functions; these are often lambdas which are a pain point for pickling. Also the scene can have updaters attached.

I agree in principle, however. After replacing the old colour module (which prohibited pickling), we also tried to replace the implementation of deep copying mobjects with pickling and unpickling -- but then hit a wall with the updaters, and haven't touched the hashing since.

Cleaning up the current serialization of mobjects would be a sane first step towards integrating multiprocessing to the scene caching.

**Reply by IAmAThousandTrees** [reply]
lamdas _can_ be pickled with dill (lol) so maybe not so hard? since 3.6 according to [SO](https://stackoverflow.com/questions/25348532/can-python-pickle-lambda-functions):

```
>> import dill
>> f=dill.loads(dill.dumps(lambda x:x*x))
>> f(5)
25
```

**Reply by IAmAThousandTrees** [reply]
also, using multiprocessing.pool and collecting the future results looks useful

**Reply by behackl** [reply]
IIRC there were some issues with dill that would have required further changes; I can't quite remember details though. Worth a try.

**Reply by IAmAThousandTrees** [reply]
further research reveals that a fork of multiprocessing that uses dill by default exists, called [multiprocess](https://github.com/uqfoundation/multiprocess/tree/master). created and actively maintained to allow distributed processing. I'll give it a try and see what the issues are, although my animations so far don't use updaters so I may not see any issues even if there are some.

**Reply by IAmAThousandTrees** [reply]
I just want to mention that I am not a reliable person to leave this to. I will probably try, but I am long-term unwell (chronic exhaustion) in a way that for some of the time affects brain function as well, making it impossible to say that I will be able to complete anything. My projects often end up on hold for months, waiting for me to have the energy to get over a bump in the road.

If someone thinks this is a useful thing to try they should take it on rather than waiting for whatever I might produce.

That said, I'm back on it today, after a break yesterday, but this is fair warning.

**Comment by lt2485923** [comment]
I second this! it would be good if the implementation of the multiprocessing was done during this great refactor. As such, these who know the codebase well can "simply" make it work with all available ressources. 

huge thanks for all the contributor!

**Comment by Flopsky** [comment]
How does the current repo benefit python3.13 updates ?
